optuna/study.py

```diff
@@ -545,6 +545,7 @@ def stop(self) -> None:
         """
 
         if self._optimize_lock.acquire(False):
+            self._optimize_lock.release()
```

Comment:
:+1:

Reply:
I put `self._optimize_lock.release()` here; otherwise, [this test](https://github.com/optuna/optuna/pull/1025/commits/7651075652e0eb67a0b61efc0f3c770d7a71c878#diff-48005dfcd962e57ca1be2ec86e03144eR751-R753) fails, where the method raises an error even inside an objective function or callback.

@toshihikoyanase 
Thanks for your additional review!